engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/event/EndEventProcessor.java

```diff
@@ -44,50 +59,147 @@ public EndEventProcessor(final BpmnBehaviors bpmnBehaviors) {
-    eventPublicationBehavior
-        .findErrorCatchEvent(errorCode, activating)
-        .ifRightOrLeft(
-            catchEvent -> {
-              stateTransitionBehavior.transitionToActivated(activating);
-              eventPublicationBehavior.throwErrorEvent(catchEvent);
-            },
-            failure -> incidentBehavior.createIncident(failure, activating));
+  @Override
+  public void onComplete(final ExecutableEndEvent element, final BpmnElementContext context) {
+    eventBehaviorOf(element).onComplete(element, context);
   }
 
   @Override
   public void onTerminate(final ExecutableEndEvent element, final BpmnElementContext terminating) {
+    eventBehaviorOf(element).onTerminate(element, terminating);
+
+    // common behavior for all end events
     incidentBehavior.resolveIncidents(terminating);
 
     final var terminated = stateTransitionBehavior.transitionToTerminated(terminating);
     stateTransitionBehavior.onElementTerminated(element, terminated);
   }
 
-  // there's some duplication here with ExclusiveGatewayProcessor where we want to short circuit and
-  // go directly from activated -> completed, which could be dry'd up
-  // TODO(npepinpe): candidate for clean up for https://github.com/camunda-cloud/zeebe/issues/6202
-  private Either<Tuple<Failure, BpmnElementContext>, BpmnElementContext> transitionUntilCompleted(
-      final ExecutableEndEvent element, final BpmnElementContext activating) {
-    if (activating.getIntent() != ProcessInstanceIntent.ELEMENT_ACTIVATING) {
-      throw new BpmnProcessingException(activating, TRANSITION_TO_COMPLETED_PRECONDITION_ERROR);
+  private EndEventBehavior eventBehaviorOf(final ExecutableEndEvent element) {
+    if (element.hasError()) {
+      return errorEndEventBehavior;
+    } else if (element.getJobWorkerProperties() != null) {
+      return messageEndEventBehavior;
+    } else {
+      return noneEndEventBehavior;
     }
+  }
+
+  /** Extract different behaviors depending on the type of event. */
+  private interface EndEventBehavior {
+    void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating);
 
-    final var activated = stateTransitionBehavior.transitionToActivated(activating);
-    final var completing = stateTransitionBehavior.transitionToCompleting(activated);
-    return stateTransitionBehavior
-        .transitionToCompleted(element, completing)
-        .mapLeft(failure -> new Tuple<>(failure, completing));
+    default void onComplete(
+        final ExecutableEndEvent element, final BpmnElementContext completing) {}
+
+    default void onTerminate(
+        final ExecutableEndEvent element, final BpmnElementContext terminating) {}
+  }
+
+  private class NoneEndEventBehavior implements EndEventBehavior {
+
+    @Override
+    public void onActivate(final ExecutableEndEvent element, final BpmnElementContext activating) {
+      transitionUntilCompleted(element, activating).ifLeft(incidentBehavior::createIncident);
```

Comment:
🔧 We could put the logic of `transactionUntilCompleted` directly into this `onActivate` method.

I reckon this was pulled out because the `onActivate` got rather large with logic when it contained both the none and the error end events. This is no longer the case thanks to your refactor 🚀